taichi/backends/opengl/opengl_api.cpp

```diff
@@ -406,8 +419,15 @@ void DeviceCompiledProgram::launch(RuntimeContext &ctx,
     // TODO: properly assert and throw if we bind more than allowed SSBOs.
     //       On most devices this number is 8. But I need to look up how
     //       to query this information so currently this is thrown from OpenGl.
-    for (const auto &[arg_id, bind_id] : program_.used.arr_arg_to_bind_idx) {
-      binder->buffer(0, bind_id, arr_bufs_[arg_id]);
+    for (const auto [arg_id, bind_id] : program_.used.arr_arg_to_bind_idx) {
+      if (!ctx.is_device_allocation[arg_id]) {
```

Comment:
nit: 
```suggestion
      if (ctx.is_device_allocation[arg_id]) {
```

then flip the two branches